feat(CLOUDDST-22843): Added rh-sign-cosign-image task

midnightercz opened the pull request:
rh-sign-cosign task allows signing snapshot via cosign signer

midnightercz committed:
feat(CLOUDDST-22843): Added rh-sign-cosign-image task

Signed-off-by: Jindrich Luza <jluza@redhat.com>

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,119 @@
+      image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
+      script: |
+        #!/usr/bin/env bash
+        if [ "$TEST" != "1" ]; then
```
The real task script should not have a special case for testing - that's why we have the mocks mechanism, so that you can override the real commands, like kubectl or cosign, with a mock. See here for an example: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-rpm-manifests-to-pyxis/tests/mocks.sh#L6

midnightercz replied:
If I understood it correctly mock script is inserted before the task script. Even if I append the mock script instead, it wouldn't help. And any more complicated modifications of the task script could cause more confusion.

mmalina replied:
Correct, mock script is typically inserted at the top. That way if the mock defines a function called `cosign`, during test execution when the task is running, the mock function is called instead of real `cosign`. See the examples I linked.

> And any more complicated modifications of the task script could cause more confusion.

The point is that you don't need to modify the task script. Instead, the mock will override all the commands that we don't want to be called in the test (using function definitions). And then the test can check afterwards that those commands were called X number of times with all the expected parameters - just like when you use mocks in python tests.

mmalina commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,20 @@
+* AWS_REGION - region of the awskms
+* AWS_ACCESS_KEY - access key for awskms
+* AWS_SECRET_ACCESS_KEY - secret access key for awskms
+* REGISTRY_USER - username for the docker registry where signatures should be stored
```
Registry access should probably just rely on an existing docker config secret bound to the service account - similar to what happens for the push-snapshot task - meaning no action is needed in the task for it to work.

midnightercz replied:
Good to know, I wasn't sure if there're already creds in existing docker/config.json

mmalina commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,20 @@
+
+Data file is expected to contain attribute sign.configMapName which is config map holding needed
+for the signing. The config map should contain the following keys:
+* SIG_KEY - key used for signing
```
All of these should probably be a secret instead of a configmap. And the secret should be mounted as a volume. Some examples:
* https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/send-slack-notification/send-slack-notification.yaml#L25
* https://github.com/hacbs-release/app-interface-deployments/blob/2e51532f2d174ce52c33a5913efe7040a5f44396/internal-services/catalog/iib-add-fbc-fragment-to-index-image-task.yaml#L273

emilyzheng replied:
Signing key is put into the same secret where other signing secrets locate in https://github.com/konflux-ci/release-service-catalog/pull/467.

mmalina replied:
Thanks. I added a new comment related to this here in the PR that will add the collect-cosign-params task: https://github.com/konflux-ci/release-service-catalog/pull/468/files#r1669971195

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,119 @@
+            # if manifest list, sign each manifest + ml itself
+            if [ $LIST -eq 1 ]; then
+              PUB_REPO_REF="${PUB_CONTAINER_REF}@${DIGEST}"
+              for MANIFEST in $(echo "$IMAGE" | jq -rc '.manifests[]'); do
```
This is iterating over an array of json strings which is potentially unsafe (in case there is a string with a space inside), so this should iterate over the array index.

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,119 @@
+                  "${INTERNAL_CONTAINER_REF}@${MDIGEST}"
+              done
+              # if sign ml itself
+              run_cosign ${COSIGN_PRE_ARGS} cosign -t 3m0s sign\
```
It seems this is exactly the same as the one below, i.e. the top level digest is always signed and then if it's a list, each item is also signed. So instead of specifying this command twice, it can just be done once outside of the `if`, like:
```sh
if [ $LIST -eq 1 ]; then
  sign each in the list
fi
sign the top level digest
```

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,119 @@
+    - name: sign-image
+      image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
+      script: |
+        #!/usr/bin/env bash
```
Ideally, the script should use `set -eux`. The `x` can be problematic because we don't want to print any secrets to the log. So it should at least use `set -eu`. But without `x`, it would be good to add some `echo` commands along the way, so that there is some useful output in the log.

mmalina replied:
There are still no echo outputs - like "processing component X" or "signing digest X with cosign" or similar.

midnightercz replied:
fixed

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,119 @@
+        COSIGN_REGISTRY_PASSWORD=$(fetch_registry_password $CONFIG_MAP_NAME)
+
+        for (( COMPONENTS_INDEX=0; COMPONENTS_INDEX<COMPONENTS_LENGTH; COMPONENTS_INDEX++ )); do
+            PUB_CONTAINER_IMAGE=$(jq -r ".components[${COMPONENTS_INDEX}].containerImage" ${SNAPSHOT_PATH})
```
Is the `PUB` prefix supposed to mean public? It's rather the opposite - the containerImage reference is the pull spec of the image that comes from the build service. `repository` is where we release it, this will be consumed by customers. So maybe `ORIGIN` or `SOURCE` would be a more appropriate prefix here. (And then `TARGET` for the other one.)

midnightercz replied:
I was using tasks/rh-sign-image/tests/test-rh-sign-image-multiple-components.yaml as source where containerImage seems to be public one. Also in
tasks/rh-sign-image/rh-sign-image.yaml quay.io
reference is replaced with registry.redhat.io

reference=$(jq -r ".components[${COMPONENTS_INDEX}].repository" ${SNAPSHOT_PATH}
reference=${reference/quay.io\/redhat-prod/registry.redhat.io}

mmalina replied:
> I was using tasks/rh-sign-image/tests/test-rh-sign-image-multiple-components.yaml as source where containerImage seems to be public one.

Yeah, unfortunately that looks like a mistake in the test. This same mistake can be seen in the push-snapshot task tests. In other places I see stuff like `"source@mydigest"`. In reality, it will be `"quay.io/redhat-user-workloads/rhtas-tenant/rekor-search-ui/rekor-search-ui@sha256:...."`

mmalina replied:
> Also in tasks/rh-sign-image/rh-sign-image.yaml quay.io reference is replaced with registry.redhat.io

Hmm, that is actually a good point. You're probably familiar with the mapping of quay repos to corresponding registry.redhat.io repos. If not, here's a brief explanation: https://github.com/konflux-ci/release-service-utils/blob/main/pyxis/create_container_image.py#L7

I can see that when you call `cosign`, you also include `--sign-container-identity`. So I wonder which identity should be used. When we sign images with simple signing using RADAS, we sign the public-facing references - registry.redhat.io and registry.access.redhat.com. Do you know if that should be the case here too? Probably, because that's what end users will use? (Customers don't have access to pull from quay.io/redhat-prod directly.)

midnightercz replied:
Yeah, sign-container-identity needs to point to public registry - probably registry.redhat.io. This identity is used in signature {"critical":{"identity":{"docker-reference": HERE. Otherwise you would have to do changes in containers policy configuration (add there quay.io)

mmalina replied:
OK, cool, so you will have to use the same approach as rh-sign-image does.

mmalina replied:
I still don't see the translation to registry.redhat.io happening anywhere.

midnightercz replied:
fixed

mmalina replied:
OK, thanks. I added another comment there.

mmalina commented on tasks/rh-sign-cosign-image/tests/pre-apply-task-hook.sh:
```diff
@@ -0,0 +1,11 @@
+
+# Add mocks to the beginning of task step script
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+yq -i '.spec.steps[0].script = load_str("'$SCRIPT_DIR'/mocks.sh") + .spec.steps[0].script' $SCRIPT_DIR/../rh-sign-cosign-image.yaml
```
The script should have `TASK_PATH="$1"` and then you can use `"$TASK_PATH"` instead of specifying the exact path to the task file. See for example here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/publish-pyxis-repository/tests/pre-apply-task-hook.sh#L3

I think we used the direct path to the task file in the past, but I cannot find an instance of this in our repo anymore. So in case you copied it from another one, please let me know which one and I can fix it there.

midnightercz committed:
pull secret args from secret
process manifest digest in manifest list directly instead of manifest
simplified code
do not pass registry-user and password to cosign

midnightercz committed:
Fixed AWS_ACCESS_KEY to AWS_ACCESS_KEY_ID
Handling component repo as public reference, containerImage as internal
Remove unneeded attributes from test configMap

midnightercz committed:
Simplified mocks

johnbieren commented on tasks/rh-sign-image/tests/test-rh-sign-image-single-component.yaml:
Are the changes to this file meant to be here? This test is for a different task that is not modified here

midnightercz replied:
This was changed by accident, I'll reverse it

johnbieren commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,20 @@
+# rh-sign-image
```
Can you update this to say `rh-sign-cosign-image`?

johnbieren commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,20 @@
+
+| Name           | Description                                                               | Optional | Default value        |
+|----------------|---------------------------------------------------------------------------|----------|----------------------|
+| snapshotPath   | Path to the JSON string of the mapped Snapshot spec in the data workspace | Yes      | snapshot_spec.json   |
```
Can you switch this to `No` for optional with no default value please?

johnbieren commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,20 @@
+| Name           | Description                                                               | Optional | Default value        |
+|----------------|---------------------------------------------------------------------------|----------|----------------------|
+| snapshotPath   | Path to the JSON string of the mapped Snapshot spec in the data workspace | Yes      | snapshot_spec.json   |
+| dataPath       | Path to the JSON string of the merged data to use in the data workspace   | Yes      | data.json            |
```
Can you switch this to `No` for optional with no default value please?

johnbieren commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+    - name: snapshotPath
+      description: Path to the JSON string of the mapped Snapshot spec in the data workspace
+      type: string
+      default: "snapshot_spec.json"
```
Can you remove this default value? We changed all of our tasks some time ago to not have default values for the *Path parameters. We found it was leading to us missing some things in our pipeline definitions

johnbieren commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+    - name: dataPath
+      description: Path to the JSON string of the merged data to use in the data workspace
+      type: string
+      default: "data.json"
```
Can you remove this default value? We changed all of our tasks some time ago to not have default values for the *Path parameters. We found it was leading to us missing some things in our pipeline definitions

johnbieren commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+metadata:
+  name: rh-sign-cosign-image
+  labels:
+    app.kubernetes.io/version: "1.1.0"
```
super minor, but we typically use `0.1.0` for the initial version of our tasks

johnbieren commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+      description: workspace to read and save files
+  steps:
+    - name: sign-image
+      image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
Can you switch this to a `konflux-ci` link instead? There was a program feature to stop having `redhat-appstudio` links in our pipelines. You should be able to find it by checking any other task definition in this repo

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-multiple-components.yaml:
The filename has to match the test name, so can you change the filename to include the `cosign` between `sign` and `image`?

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,134 @@
+          - name: data
+        steps:
+          - name: setup-values
+            image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
same comment about the image as in the task

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,134 @@
+      taskSpec:
+        steps:
+          - name: check-result
+            image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
same image comment

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,134 @@
+          - name: data
+        steps:
+          - name: delete-crs
+            image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
same image comment

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,134 @@
+              [ "$CALLS" = "$EXPECTED" ] || exit 1
+      runAfter:
+        - run-task
+  finally:
```
I don't think we need this task in the test. The workspace is just a volume claim that will be released after the pipeline finishes IIUC. At least, we don't clear out the workspace at the end of our other test files (we do remove the CRs, but no CRs are created here)

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-single-component.yaml:
Needs `cosign` between `sign` and `image` in filename

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-single-component.yaml:
```diff
@@ -0,0 +1,104 @@
+          - name: data
+        steps:
+          - name: setup-values
+            image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
same image comment

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-single-component.yaml:
```diff
@@ -0,0 +1,104 @@
+      taskSpec:
+        steps:
+          - name: check-result
+            image: quay.io/redhat-appstudio/release-service-utils:447ea0580a2cdd48b4091e1df86fab5c3f86d01c
```
same image comment

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-image-single-component.yaml:
```diff
@@ -0,0 +1,104 @@
+              [ "$CALLS" = "$EXPECTED" ] || exit 1
+      runAfter:
+        - run-task
+  finally:
```
Don't think you need from here down (same comment as other test)

midnightercz replied:
When I run tests locally it keeps workspace for both test cases and therefore mock calls files contain records from previous test

mmalina replied:
Are you sure? I don't know how that could be the case, since, as Johnny said, each pipeline run is tarted with its own PVC: https://github.com/konflux-ci/release-service-catalog/blob/development/.github/scripts/test_tekton_tasks.sh#L119
```sh
    PIPELINERUN=$(tkn p start $TEST_NAME -w name=tests-workspace,volumeClaimTemplateFile=$WORKSPACE_TEMPLATE -o json | jq -r '.metadata.name')
```

The template is here: https://github.com/konflux-ci/release-service-catalog/blob/development/.github/resources/workspace-template.yaml

midnightercz replied:
yeah, you're right, I guess it was before rebase

johnbieren reviewed (commented):
Added some comments - thank you for adjusting the PR to our style. I find it much easier to understand now

midnightercz committed:
Updated README
Chnaged test filenames
No default values for params

mmalina commented:
@midnightercz you asked us to review again, but I don't know what to review since you resolved all the threads. So now I will have to expand all the resolved threads and check each one even if some of them may be no-op. So for next time please leave resolving of threads to the person that opened the thread and instead please add a comment to saying what was done - often "Fixed" is enough, but sometimes there are multiple options, so then it's good to say what was done.

mmalina commented:
Also, please check the yamllint and gitlint issues.

midnightercz commented:
@mmalina I pinged you to merge this if you think it's ready to merge or find another issues. I'll check pylint error

midnightercz committed:
Fixed yamllint issues

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,103 @@
+        #!/usr/bin/env bash
+        set -eu
+        function fetch_cosign_key() {
+          echo $(kubectl_get_arg_from_config_map $1 SIG_KEY)
```
Not sure how this is supposed to work. `kubectl_get_arg_from_config_map` is undefined here (it's only defined in context of tests via mocks.sh).

midnightercz replied:
Right, I forgot to change this one

midnightercz replied:
fixed

mmalina replied:
I added a new comment there.

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,103 @@
+            if [ $LIST -eq 1 ]; then
+              PUB_REPO_REF="${PUB_CONTAINER_REF}@${DIGEST}"
+              for MDIGEST in $(echo "$IMAGE" | jq -rc '.manifests[]|.digest'); do
+                cosign ${COSIGN_PRE_ARGS} cosign -t 3m0s sign\
```
This looks like a typo. I don't think `cosign cosign sign` is the correct command - should it be just `cosign sign`? https://github.com/sigstore/cosign/blob/main/doc/cosign_sign.md

And also, I'm curious about the `COSIGN_PRE_ARGS` variable. It will contain `AWS_REGION=...` etc. But these are not valid `cosign` argument, are they? Or are they meant to be env vars supplied to `cosign` (that would mean that the leading `cosign` shouldn't be there)? In that case it would be redundant, because they supplied to everything that is invoked in this script

midnightercz replied:
yeah, there's typo, COSIGN_PRE_ARGS are env variables needed for cosign (if you're using AWSKMS). So in practice envs are not needed to be set explicitly, but it just feels better to explicitly put in here to emphasize they are needed. But I can remove them if you insist

mmalina replied:
One argument for defining env vars explicitly for a command is that in principal it's safer if the env vars are only available to the command that needs it. But this is not the case here anyway - here it just seems redundant since the env vars are already there. So yeah, I would prefer for them to be removed.

midnightercz replied:
fixed

mmalina replied:
Thanks.

mmalina commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,122 @@
+            script: |
+              #!/usr/bin/env sh
+              set -eux
+              # empty mock calls files
```
What's this comment? A leftover from something that was removed?

midnightercz replied:
fixed

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,103 @@
+
+        COMPONENTS_LENGTH=$(jq '.components |length' ${SNAPSHOT_PATH})
+
+        COSIGN_SIG_KEY=$(fetch_cosign_key $CONFIG_MAP_NAME)
```
Again, I'm not sure what this is supposed to do (it currently does nothing), but since it's a secret, it should probably use a secret and not a configmap. But looking at cosign docs: https://github.com/sigstore/cosign/blob/main/doc/cosign_sign.md

> path to the private key file, KMS URI or Kubernetes Secret

So maybe you can pass a secret name directly and it will fetch it? Not sure what the expected key in the secret would be though. Otherwise a file can be prepared.

midnightercz replied:
key argument itself is not secret, in our use case it will be something like awskms://.... so as it's not secret it should be in configmap. I'm not sure how has access to see secrets in ocp, but I assume relengs should be able to see what signing key is used to sign their products

mmalina replied:
I see. So in this case it's some kind of aws path that is used to fetch the actual key. Then it sounds good.

mmalina commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,122 @@
+              EOF
+              )
+              CALLS=$(cat "$(workspaces.data.path)/mock_skopeo_calls")
+              [ "$CALLS" = "$EXPECTED" ] || exit 1
```
The `|| exit 1` is redundant here. Also, I would suggest running it as `test "$CALLS" = "$EXPECTED"`. It does the same thing, but it makes it consistent with our other tests and it makes the tests easy to spot in the check-results script.

midnightercz replied:
Yeah, I usually do && || so I put it there just as a habit. I'll change it to test

midnightercz replied:
fixed

mmalina replied:
Cool, thanks.

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,103 @@
+                  "${INTERNAL_CONTAINER_REF}@${MDIGEST}"
+              done
+            fi
+              # if sign ml itself or just manifest (if it's not list)
+              cosign ${COSIGN_PRE_ARGS} cosign -t 3m0s sign\
+                ${COSIGN_COMMON_ARGS}\
+                --sign-container-identity "${PUB_CONTAINER_REF}@${DIGEST}"\
+                "${INTERNAL_CONTAINER_REF}@${DIGEST}"
```
The indentation here should be 2 spaces less.

midnightercz replied:
fixed

mmalina replied:
I still see wrong indentation:
```sh
            fi
              # if sign ml itself or just manifest (if it's not list)
              echo "Signing manifest ${INTERNAL_CONTAINER_REF}@${DIGEST}"
              cosign -t 3m0s sign\
              ${COSIGN_COMMON_ARGS}\
              --sign-container-identity "${PUB_CONTAINER_REF}@${DIGEST}"\
              "${INTERNAL_CONTAINER_REF}@${DIGEST}"
            echo "done"
```
Also, the comment "if sign..." - probably the "if" should be dropped, because now there's no if:
So should be:
```sh
            fi
            # sign ml itself or just manifest (if it's not list)
            echo "Signing manifest ${INTERNAL_CONTAINER_REF}@${DIGEST}"
            cosign -t 3m0s sign\
              ${COSIGN_COMMON_ARGS}\
              --sign-container-identity "${PUB_CONTAINER_REF}@${DIGEST}"\
              "${INTERNAL_CONTAINER_REF}@${DIGEST}"
            echo "done"
```

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

mmalina replied:
Thanks.

mmalina reviewed (commented):
Added a few more comments and reopened some threads.

midnightercz committed:
Fixed public/internal reference
Added info echos
Remove cosign pre args

johnbieren reviewed (commented):
I don't have anything to ask to change beyond what Martin already has open threads on :slightly_smiling_face:

midnightercz committed:
Changed [ ] to test and removed redundant 'exit 1'

mmalina commented on tasks/push-to-cdn/README.md:
This change shouldn't really be here, right? Can you rebase or something?

midnightercz replied:
That's what happened when I did the rebase. I'll try to figure out how to get rid of that commit

midnightercz replied:
ok rebase and force push helped

mmalina replied:
Yeah, if you use the rebase button in github, I think it often messes things up. It's best to just do a local rebase and then force push, now it's ok, thanks.

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+            echo "Processing component ${COMPONENT_NAME}"
+            INTERNAL_CONTAINER_REF=$(jq -r ".components[${COMPONENTS_INDEX}].repository" ${SNAPSHOT_PATH})
+            INTERNAL_CONTAINER_IMAGE="${INTERNAL_CONTAINER_REF}@${DIGEST}"
+            PUB_CONTAINER_REF=$(translate-delivery-repo $INTERNAL_CONTAINER_REF | jq -rc ".[0]|.url" | tr -d \")
```
The `tr` shoudln't be needed, right? Because you use `jq -r`, it will be the raw string without `""`. Also, the `-c` option is redundant in the `jq` - it's just a string, so it cannot be made any more compact. Also, the `|` is redundant as well. So I would suggest this:
```sh
PUB_CONTAINER_REF=$(translate-delivery-repo $INTERNAL_CONTAINER_REF | jq -r ".[0].url")
```

mmalina replied:
Also, I was looking at what we do in rh-sign-image and I can see we sign the registry.access.redhat.com reference as well. I'm not even sure we could sign multiple times with cosign. And it's probably ok to just use the registry.redhat.io reference. But I have another question - in this case we only sign the container image referenced by digest, is that correct? Compare this to simple signing (rh-sign-image) where we sign every tag separately. I think the answer is that here we do want to sign just the digest reference. But I am not 100 % sure.

midnightercz replied:
> But I have another question - in this case we only sign the container image referenced by digest, is that correct? 

Yes, that's correct. Cosign has optional annotation where you can specify tag, but it's used only as optional attribute in the signature.
I wasn't sure how to identity works as it wasn't present in cosign before. But it looks like when you sign same digest with another identity it will just create another simplesigning.v1+json cosign blob. So code can be changed to sign multiple times for different registries

mmalina replied:
Thanks. I can see this was done - at least when reviewing the follow up PR: https://github.com/konflux-ci/release-service-catalog/pull/467

mmalina commented on tasks/rh-sign-cosign-image/rh-sign-cosign-image.yaml:
```diff
@@ -0,0 +1,104 @@
+        #!/usr/bin/env bash
+        set -eu
+        function fetch_cosign_key() {
+          kubectl get configmap --output json | jq -r ".data.SIG_KEY" | tr -d \"
```
I don't think this will work. The name of the configmap to fetch is missing - I guess it should be the one from the data json. But overall I'm still not sure about this - I think this should be a secret. And because of this (and the aws stuff) I think we'll also need a "prepare-cosign-sign-parameters" task like we have for pyxis stuff and now also for the cdn tasks that Scott W. was working on - @johnbieren  you know what I'm talking about, right? WDYT?

(Also, the `tr` seems redundant here.)

emilyzheng replied:
Changed in https://github.com/konflux-ci/release-service-catalog/pull/467. Key is got from secret now.

mmalina replied:
Thanks.

midnightercz committed:
Removed tr -d
Support multiple identities

midnightercz commented:
I pushed change to support multiple identities. But I'm starting my parental leave on Monday and will be back on 2 weeks. If there are any critical changes please feel free to submit them by yourselves

johnbieren commented on tasks/rh-sign-cosign-image/README.md:
```diff
@@ -0,0 +1,16 @@
+# rh-sign-cosign-image
+
+Task to create sign containers in snapshot with cosign
```
I think the word create in this line was accidentally leftover

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

mmalina replied:
Thanks.

johnbieren commented on tasks/rh-sign-cosign-image/tests/pre-apply-task-hook.sh:
```diff
@@ -0,0 +1,20 @@
+# Add RBAC so that the SA executing the tests can retrieve CRs
+kubectl apply -f .github/resources/crd_rbac.yaml
+
+# Create a dummy slack-notification-secret secret (and delete it first if it exists)
```
`s/slack-notification/rh-sign-cosign/`

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

mmalina replied:
Thanks.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+              #!/usr/bin/env sh
+              set -eux
+              _TEST_MANIFEST_LIST_DIGEST="sha256:1111"
+              _TEST_MANIFEST_LIST_MDIGEST1="sha256:1111-1"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+              set -eux
+              _TEST_MANIFEST_LIST_DIGEST="sha256:1111"
+              _TEST_MANIFEST_LIST_MDIGEST1="sha256:1111-1"
+              _TEST_MANIFEST_LIST_MDIGEST2="sha256:1111-2"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+              _TEST_MANIFEST_LIST_DIGEST="sha256:1111"
+              _TEST_MANIFEST_LIST_MDIGEST1="sha256:1111-1"
+              _TEST_MANIFEST_LIST_MDIGEST2="sha256:1111-2"
+              _TEST_MANIFEST_LIST_MDIGEST3="sha256:1111-3"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+              _TEST_MANIFEST_LIST_REPO="quay.io/redhat-pending/test-product----test-image0"
+
+              _TEST_MANIFEST_LIST_OCI_DIGEST="sha256:0000"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST1="sha256:0000-1"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+
+              _TEST_MANIFEST_LIST_OCI_DIGEST="sha256:0000"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST1="sha256:0000-1"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST2="sha256:0000-2"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

johnbieren commented on tasks/rh-sign-cosign-image/tests/test-rh-sign-cosign-image-multiple-components.yaml:
```diff
@@ -0,0 +1,130 @@
+              _TEST_MANIFEST_LIST_OCI_DIGEST="sha256:0000"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST1="sha256:0000-1"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST2="sha256:0000-2"
+              _TEST_MANIFEST_LIST_OCI_MDIGEST3="sha256:0000-3"
```
I don't see this used

emilyzheng replied:
Fixed in https://github.com/konflux-ci/release-service-catalog/pull/467.

mmalina reviewed (commented):
I closed all the remaining threads. There is now a new PR that addresses all the threads that were remaining: https://github.com/konflux-ci/release-service-catalog/pull/467

As I understand it, the new PR was created because Emily took over this task for now and she likely didn't have permissions to push to Jindra's branch to continue here.

So this PR will be closed in favour of the new one.

johnbieren commented:
Closing per the last comment. Feel free to reopen if I should not have closed this
